R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
+# When appId is provided, all other information is secondary. An appId is an indication from the
+# caller that the content has already been deployed elsewhere. If we cannot locate that content,
+# deployment fails.
+#
+# The target content may have been created by some other user; the account for this session may
+# differ from the account used when creating the content.
+deploymentTargetFromAppId <- function(
+  recordPath = ".",
+  appId = NULL,
+  appName = NULL,
+  appTitle = NULL,
+  envVars = NULL,
+  account = NULL,
+  server = NULL,
+  error_call = caller_env()
+) {
 
-    createDeploymentTarget(
-      appName = appName,
-      appTitle = appTitle,
-      appId = appId,
-      envVars = envVars,
-      username = fullAccount$name, # first deploy must be to own account
-      account = fullAccount$name,
-      server = fullAccount$server
+  # We must have a target account+server in order to use the appId.
+  # The selected account may not be the original creator of the content.
+  accountDetails <- accountInfo(account, server)
+
+  # Filtering is only by server and includes all deployments in case we have a deployment record
+  # from a collaborator.
+  appDeployments <- deployments(
+    appPath = recordPath,
+    serverFilter = server,
+    excludeOrphaned = FALSE
+  )
+  appDeployments <- appDeployments[appDeployments$appId == appId, ]
+  if (nrow(appDeployments) > 1) {
+    cli::cli_abort(
+      c(
+        "Supplied {.arg appId} ({appId}) identifies multiple deployments.",
+        i = "Manage obsolete deployments with rsconnect::forgetDeployment()."
+      ),
+      call = error_call
     )
-  } else if (nrow(appDeployments) == 1) {
-    # If both appName and appId supplied, check that they're consistent.
-    if (!is.null(appId) && appDeployments$appId != appId) {
-      cli::cli_abort(
-        c(
-          "Supplied {.arg appId} ({appId}) does not match deployment record ({appDeployments$appId}).",
-          i = "Omit {.arg appId} to use existing for deployment for app {.str {appName}}, or",
-          i = "Omit {.arg appName} to create new deployment record."
-        ),
-        call = error_call
-      )
-    }
+  }
 
-    updateDeploymentTarget(appDeployments, appTitle, envVars)
-  } else {
-    deployment <- disambiguateDeployments(appDeployments, error_call = error_call)
-    updateDeploymentTarget(deployment, appTitle, envVars)
+  # Existing local deployment record.
+  if (nrow(appDeployments) == 1) {
+    deployment <- appDeployments[1, ]
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
   }
-}
 
-deploymentTargetForApp <- function(appId,
-                                   appTitle = NULL,
-                                   account = NULL,
-                                   server = NULL) {
-  accountDetails <- findAccount(account, server)
+  # No local deployment record. Get it from the server.
```

Comment:
I think this would be more clear if you three if clauses corresponding to 0 matches, 1 match, more than 1 match.

Reply:
I'm not sure I follow. We have:

```r
if (num > 1) {
  error
}

if (num == 1) {
  return(...)
}

# else no matches. do some work then
return(...)
```

I tend to favor this style because it avoids deep conditional clauses in many scenarios. What is the advantage of forcing this into an explicit if-else-if-else?

Reply:
I like the separate if statements when the code blocks are long, but here where they're relatively short, I think the single if block makes it easier to verify that you've covered all the conditions:

```R
if (num > 1) {

}  else if (num == 1) {

} else } # num == 0

}
```

Reply:
I'm going to leave this as-is for now, as each of the three `findDeploymentTarget*` functions follow the same pattern -- dispatch for some cases then do something slightly more complicated when there are no records.